File: examples/smw_helpers_demo.js
/**
 * SMW Helpers Demonstration
 * 
 * Shows how to use the SMW helper library to manipulate
 * Super Mario World game state
 * 
 * Prerequisites:
 * - QUsb2snes or USB2SNES server running
 * - SMW ROM loaded on console
 * - In a level (for most functions to work)
 */

const { SNESWrapper } = require('../electron/main/usb2snes/SNESWrapper');
const SMWHelpers = require('../electron/main/smw/SMWHelpers');

async function demonstrateSMWHelpers() {
  console.log('=== SMW Helpers Demonstration ===\n');
  
  // Connect to console
  console.log('Connecting to USB2SNES...');
  const snes = new SNESWrapper();
  
  try {
    await snes.fullConnect('usb2snes_a', 'ws://localhost:64213');
    console.log('✓ Connected!\n');
  } catch (error) {
    console.error('✗ Connection failed:', error.message);
    console.log('\nMake sure:');
    console.log('  1. QUsb2snes is running');
    console.log('  2. Console is powered on');
    console.log('  3. SMW ROM is loaded');
    process.exit(1);
  }
  
  // Create SMW helpers
  const smw = new SMWHelpers(snes);
  
  // ========================================
  // Demo 1: Read Current Game State
  // ========================================
  console.log('--- Demo 1: Current Game State ---');
  
  const mode = await smw.getGameMode();
  console.log(`Game Mode: 0x${mode.toString(16).toUpperCase()}`);
  
  if (mode === smw.MODES.LEVEL) {
    console.log('  → In Level ✓');
  } else if (mode === smw.MODES.OVERWORLD) {
    console.log('  → On Overworld');
  } else if (mode === smw.MODES.TITLE) {
    console.log('  → At Title Screen');
  } else {
    console.log(`  → Other mode (${mode})`);
  }
  
  const lives = await smw.getLives();
  const coins = await smw.getCoins();
  const powerup = await smw.getPowerup();
  
  console.log(`Lives: ${lives}`);
  console.log(`Coins: ${coins}`);
  console.log(`Powerup: ${smw.getPowerupName(powerup)} (${powerup})`);
  
  const hasYoshi = await smw.hasYoshi();
  if (hasYoshi) {
    const yoshiColor = await smw.getYoshiColor();
    const colorNames = ['Green', 'Red', 'Blue', 'Yellow'];
    console.log(`Yoshi: ${colorNames[yoshiColor]} ✓`);
  } else {
    console.log('Yoshi: None');
  }
  
  console.log();
  
  // ========================================
  // Demo 2: Batch Read (Fast!)
  // ========================================
  console.log('--- Demo 2: Batch Read (Fast!) ---');
  
  const startTime = Date.now();
  const state = await smw.getGameState();
  const elapsed = Date.now() - startTime;
  
  console.log(`Got entire game state in ${elapsed}ms:`);
  console.log(`  Lives: ${state.lives}`);
  console.log(`  Coins: ${state.coins}`);
  console.log(`  Powerup: ${smw.getPowerupName(state.powerup)}`);
  console.log(`  Position: (${state.position.x}, ${state.position.y})`);
  console.log(`  Yoshi: ${state.hasYoshi ? 'Yes' : 'No'}`);
  console.log(`  Sprites Locked: ${state.spritesLocked ? 'Yes' : 'No'}`);
  console.log();
  
  // ========================================
  // Demo 3: Modify Game State
  // ========================================
  console.log('--- Demo 3: Modify Game State ---');
  console.log('WARNING: This will change your game state!');
  console.log('Press Ctrl+C to cancel, or wait 3 seconds...\n');
  
  await sleep(3000);
  
  console.log('Setting lives to 99...');
  await smw.setLives(99);
  const newLives = await smw.getLives();
  console.log(`✓ Lives now: ${newLives}`);
  
  console.log('Setting coins to 50...');
  await smw.setCoins(50);
  const newCoins = await smw.getCoins();
  console.log(`✓ Coins now: ${newCoins}`);
  
  if (await smw.isInLevel()) {
    console.log('Giving cape powerup...');
    await smw.setPowerup(smw.POWERUPS.CAPE);
    const newPowerup = await smw.getPowerup();
    console.log(`✓ Powerup now: ${smw.getPowerupName(newPowerup)}`);
    
    if (!await smw.hasYoshi()) {
      console.log('Giving green Yoshi...');
      await smw.giveYoshi(smw.YOSHI.GREEN);
      console.log('✓ Yoshi given!');
    }
  } else {
    console.log('(Skipping powerup/Yoshi - not in level)');
  }
  
  console.log();
  
  // ========================================
  // Demo 4: Position Queries (In Level Only)
  // ========================================
  if (await smw.isInLevel()) {
    console.log('--- Demo 4: Position Queries ---');
    
    const pos = await smw.getPosition();
    console.log(`Current Position: (${pos.x}, ${pos.y})`);
    
    const speed = await smw.getSpeed();
    console.log(`Current Speed: X=${speed.x}, Y=${speed.y}`);
    
    const dir = await smw.getDirection();
    console.log(`Direction: ${dir === 0 ? 'Right' : 'Left'}`);
    
    const isFlying = await smw.isFlying();
    const isDucking = await smw.isDucking();
    const isSwimming = await smw.isSwimming();
    
    console.log(`Flying: ${isFlying ? 'Yes' : 'No'}`);
    console.log(`Ducking: ${isDucking ? 'Yes' : 'No'}`);
    console.log(`Swimming: ${isSwimming ? 'Yes' : 'No'}`);
    console.log();
  }
  
  // ========================================
  // Demo 5: Level Queries
  // ========================================
  if (await smw.isInLevel()) {
    console.log('--- Demo 5: Level Queries ---');
    
    const isVertical = await smw.isVerticalLevel();
    const isWater = await smw.isWaterLevel();
    
    console.log(`Vertical Level: ${isVertical ? 'Yes' : 'No'}`);
    console.log(`Water Level: ${isWater ? 'Yes' : 'No'}`);
    console.log();
  }
  
  // ========================================
  // Demo 6: Special Items
  // ========================================
  if (await smw.isInLevel()) {
    console.log('--- Demo 6: Special Items ---');
    
    const pTimer = await smw.getBluePowTimer();
    console.log(`P-Switch Timer: ${pTimer} frames`);
    
    const onOffStatus = await smw.getOnOffStatus();
    console.log(`ON/OFF Switch: ${onOffStatus ? 'Yellow Outline' : 'Yellow Blocks'}`);
    
    console.log('\nActivating P-switch for 5 seconds...');
    await smw.activatePSwitch(300);  // 300 frames = 5 seconds
    console.log('✓ P-switch activated!');
    console.log('  (Watch yellow blocks turn into coins!)');
    console.log();
  }
  
  // ========================================
  // Demo 7: Sprite Control
  // ========================================
  if (await smw.isInLevel()) {
    console.log('--- Demo 7: Sprite Control ---');
    
    const spritesLocked = await smw.areSpritesLocked();
    console.log(`Sprites Currently: ${spritesLocked ? 'Frozen' : 'Active'}`);
    
    console.log('Freezing all sprites for 3 seconds...');
    await smw.setSpritesLocked(true);
    console.log('✓ Sprites frozen! (Enemies stopped moving)');
    
    await sleep(3000);
    
    console.log('Unfreezing sprites...');
    await smw.setSpritesLocked(false);
    console.log('✓ Sprites unfrozen!');
    console.log();
  }
  
  // ========================================
  // Demo 8: Memory Watcher
  // ========================================
  console.log('--- Demo 8: Memory Watcher ---');
  console.log('Watching for state changes for 10 seconds...');
  console.log('(Try collecting coins, changing powerup, etc.)\n');
  
  let changeCount = 0;
  const watcher = smw.createStateWatcher((changes) => {
    changeCount += changes.length;
    changes.forEach(change => {
      const addr = change.address;
      if (addr === smw.RAM.StatusLives) {
        console.log(`  → Lives changed: ${change.oldValue[0]} → ${change.newValue[0]}`);
      } else if (addr === smw.RAM.StatusCoins) {
        console.log(`  → Coins changed: ${change.oldValue[0]} → ${change.newValue[0]}`);
      } else if (addr === smw.RAM.MarioPowerUp) {
        console.log(`  → Powerup changed: ${smw.getPowerupName(change.oldValue[0])} → ${smw.getPowerupName(change.newValue[0])}`);
      } else if (addr === smw.RAM.OnYoshi) {
        console.log(`  → Yoshi: ${change.oldValue[0] ? 'On' : 'Off'} → ${change.newValue[0] ? 'On' : 'Off'}`);
      }
    });
  }, 100);  // Poll every 100ms
  
  await watcher.start();
  await sleep(10000);
  watcher.stop();
  
  console.log(`\nDetected ${changeCount} state changes in 10 seconds.`);
  console.log();
  
  // ========================================
  // Summary
  // ========================================
  console.log('=== Demo Complete! ===');
  console.log('\nSMW Helpers provides 40+ functions:');
  console.log('  ✓ Player state (lives, coins, powerup, position, speed)');
  console.log('  ✓ Yoshi control (give, remove, colors, wings)');
  console.log('  ✓ Sprite control (freeze, kill, query)');
  console.log('  ✓ Game queries (mode, level type, flags)');
  console.log('  ✓ Special items (P-switch, ON/OFF)');
  console.log('  ✓ Memory watchers (auto-detect changes)');
  console.log('  ✓ Batch operations (3x faster!)');
  console.log('\nSee devdocs/SMW_HELPERS_GUIDE.md for complete API reference.');
  
  process.exit(0);
}

function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

// Run demonstration
demonstrateSMWHelpers().catch(error => {
  console.error('\n✗ Error:', error.message);
  console.error(error.stack);
  process.exit(1);
});

File: examples/smw_helpers_demo.py
"""
SMW Helpers Demonstration (Python)

Shows how to use the SMW helper library to manipulate
Super Mario World game state

Prerequisites:
- QUsb2snes or USB2SNES server running
- SMW ROM loaded on console
- In a level (for most functions to work)
"""

import asyncio
import sys
import os

# Add parent directory to path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..'))

from py2snes import py2snes
from py2snes.smw import SMWHelpers, GAME_MODES, POWERUPS, YOSHI_COLORS

async def demonstrate_smw_helpers():
    print('=== SMW Helpers Demonstration (Python) ===\n')
    
    # Connect to console
    print('Connecting to USB2SNES...')
    snes = py2snes.snes()
    
    try:
        await snes.connect('ws://localhost:64213')
        devices = await snes.DeviceList()
        if not devices:
            raise Exception('No devices found')
        await snes.Attach(devices[0])
        print('✓ Connected!\n')
    except Exception as error:
        print(f'✗ Connection failed: {error}')
        print('\nMake sure:')
        print('  1. QUsb2snes is running')
        print('  2. Console is powered on')
        print('  3. SMW ROM is loaded')
        sys.exit(1)
    
    # Create SMW helpers
    smw = SMWHelpers(snes)
    
    # ========================================
    # Demo 1: Read Current Game State
    # ========================================
    print('--- Demo 1: Current Game State ---')
    
    mode = await smw.get_game_mode()
    print(f'Game Mode: 0x{mode:02X}')
    
    if mode == GAME_MODES.LEVEL:
        print('  → In Level ✓')
    elif mode == GAME_MODES.OVERWORLD:
        print('  → On Overworld')
    elif mode == GAME_MODES.TITLE:
        print('  → At Title Screen')
    else:
        print(f'  → Other mode ({mode})')
    
    lives = await smw.get_lives()
    coins = await smw.get_coins()
    powerup = await smw.get_powerup()
    
    print(f'Lives: {lives}')
    print(f'Coins: {coins}')
    print(f'Powerup: {smw.get_powerup_name(powerup)} ({powerup})')
    
    has_yoshi = await smw.has_yoshi()
    if has_yoshi:
        yoshi_color = await smw.get_yoshi_color()
        color_names = ['Green', 'Red', 'Blue', 'Yellow']
        print(f'Yoshi: {color_names[yoshi_color]} ✓')
    else:
        print('Yoshi: None')
    
    print()
    
    # ========================================
    # Demo 2: Batch Read (Fast!)
    # ========================================
    print('--- Demo 2: Batch Read (Fast!) ---')
    
    import time
    start_time = time.time()
    state = await smw.get_game_state()
    elapsed = (time.time() - start_time) * 1000
    
    print(f'Got entire game state in {elapsed:.1f}ms:')
    print(f'  Lives: {state["lives"]}')
    print(f'  Coins: {state["coins"]}')
    print(f'  Powerup: {smw.get_powerup_name(state["powerup"])}')
    print(f'  Position: ({state["position"]["x"]}, {state["position"]["y"]})')
    print(f'  Yoshi: {"Yes" if state["has_yoshi"] else "No"}')
    print(f'  Sprites Locked: {"Yes" if state["sprites_locked"] else "No"}')
    print()
    
    # ========================================
    # Demo 3: Modify Game State
    # ========================================
    print('--- Demo 3: Modify Game State ---')
    print('WARNING: This will change your game state!')
    print('Press Ctrl+C to cancel, or wait 3 seconds...\n')
    
    await asyncio.sleep(3)
    
    print('Setting lives to 99...')
    await smw.set_lives(99)
    new_lives = await smw.get_lives()
    print(f'✓ Lives now: {new_lives}')
    
    print('Setting coins to 50...')
    await smw.set_coins(50)
    new_coins = await smw.get_coins()
    print(f'✓ Coins now: {new_coins}')
    
    if await smw.is_in_level():
        print('Giving cape powerup...')
        await smw.set_powerup(POWERUPS.CAPE)
        new_powerup = await smw.get_powerup()
        print(f'✓ Powerup now: {smw.get_powerup_name(new_powerup)}')
        
        if not await smw.has_yoshi():
            print('Giving green Yoshi...')
            await smw.give_yoshi(YOSHI_COLORS.GREEN)
            print('✓ Yoshi given!')
    else:
        print('(Skipping powerup/Yoshi - not in level)')
    
    print()
    
    # ========================================
    # Demo 4: Position Queries (In Level Only)
    # ========================================
    if await smw.is_in_level():
        print('--- Demo 4: Position Queries ---')
        
        pos = await smw.get_position()
        print(f'Current Position: ({pos["x"]}, {pos["y"]})')
        
        speed = await smw.get_speed()
        print(f'Current Speed: X={speed["x"]}, Y={speed["y"]}')
        
        direction = await smw.get_direction()
        print(f'Direction: {"Right" if direction == 0 else "Left"}')
        
        is_flying = await smw.is_flying()
        is_ducking = await smw.is_ducking()
        is_swimming = await smw.is_swimming()
        
        print(f'Flying: {"Yes" if is_flying else "No"}')
        print(f'Ducking: {"Yes" if is_ducking else "No"}')
        print(f'Swimming: {"Yes" if is_swimming else "No"}')
        print()
    
    # ========================================
    # Demo 5: Level Queries
    # ========================================
    if await smw.is_in_level():
        print('--- Demo 5: Level Queries ---')
        
        is_vertical = await smw.is_vertical_level()
        is_water = await smw.is_water_level()
        
        print(f'Vertical Level: {"Yes" if is_vertical else "No"}')
        print(f'Water Level: {"Yes" if is_water else "No"}')
        print()
    
    # ========================================
    # Demo 6: Special Items
    # ========================================
    if await smw.is_in_level():
        print('--- Demo 6: Special Items ---')
        
        p_timer = await smw.get_blue_pow_timer()
        print(f'P-Switch Timer: {p_timer} frames')
        
        on_off_status = await smw.get_on_off_status()
        print(f'ON/OFF Switch: {"Yellow Outline" if on_off_status else "Yellow Blocks"}')
        
        print('\nActivating P-switch for 5 seconds...')
        await smw.activate_p_switch(300)  # 300 frames = 5 seconds
        print('✓ P-switch activated!')
        print('  (Watch yellow blocks turn into coins!)')
        print()
    
    # ========================================
    # Demo 7: Sprite Control
    # ========================================
    if await smw.is_in_level():
        print('--- Demo 7: Sprite Control ---')
        
        sprites_locked = await smw.are_sprites_locked()
        print(f'Sprites Currently: {"Frozen" if sprites_locked else "Active"}')
        
        print('Freezing all sprites for 3 seconds...')
        await smw.set_sprites_locked(True)
        print('✓ Sprites frozen! (Enemies stopped moving)')
        
        await asyncio.sleep(3)
        
        print('Unfreezing sprites...')
        await smw.set_sprites_locked(False)
        print('✓ Sprites unfrozen!')
        print()
    
    # ========================================
    # Demo 8: Memory Watcher
    # ========================================
    print('--- Demo 8: Memory Watcher ---')
    print('Watching for state changes for 10 seconds...')
    print('(Try collecting coins, changing powerup, etc.)\n')
    
    change_count = [0]  # Use list for mutable reference
    
    def on_change(changes):
        change_count[0] += len(changes)
        for change in changes:
            addr = change['address']
            if addr == smw.RAM.StatusLives:
                print(f'  → Lives changed: {change["old_value"][0]} → {change["new_value"][0]}')
            elif addr == smw.RAM.StatusCoins:
                print(f'  → Coins changed: {change["old_value"][0]} → {change["new_value"][0]}')
            elif addr == smw.RAM.MarioPowerUp:
                print(f'  → Powerup changed: {smw.get_powerup_name(change["old_value"][0])} → {smw.get_powerup_name(change["new_value"][0])}')
            elif addr == smw.RAM.OnYoshi:
                print(f'  → Yoshi: {"On" if change["old_value"][0] else "Off"} → {"On" if change["new_value"][0] else "Off"}')
    
    watcher = smw.create_state_watcher(on_change, 0.1)  # Poll every 100ms
    
    await watcher.start()
    await asyncio.sleep(10)
    watcher.stop()
    
    print(f'\nDetected {change_count[0]} state changes in 10 seconds.')
    print()
    
    # ========================================
    # Summary
    # ========================================
    print('=== Demo Complete! ===')
    print('\nSMW Helpers provides 40+ functions:')
    print('  ✓ Player state (lives, coins, powerup, position, speed)')
    print('  ✓ Yoshi control (give, remove, colors, wings)')
    print('  ✓ Sprite control (freeze, kill, query)')
    print('  ✓ Game queries (mode, level type, flags)')
    print('  ✓ Special items (P-switch, ON/OFF)')
    print('  ✓ Memory watchers (auto-detect changes)')
    print('  ✓ Batch operations (3x faster!)')
    print('\nSee devdocs/SMW_HELPERS_GUIDE.md for complete API reference.')

# Run demonstration
if __name__ == '__main__':
    try:
        asyncio.run(demonstrate_smw_helpers())
    except KeyboardInterrupt:
        print('\n\nDemo interrupted by user.')
    except Exception as error:
        print(f'\n✗ Error: {error}')
        import traceback
        traceback.print_exc()
        sys.exit(1)

